File: dotnet/Library/Okapi/Oberon.cs
using Okapi.Security.V1;

namespace Okapi.Security;

public static class Oberon
{
    public static CreateOberonKeyResponse CreateKey(CreateOberonKeyRequest request)
    {
        return Native.Call<CreateOberonKeyRequest, CreateOberonKeyResponse>(request, Native.oberon_create_key);
    }

    public static CreateOberonTokenResponse CreateToken(CreateOberonTokenRequest request)
    {
        return Native.Call<CreateOberonTokenRequest, CreateOberonTokenResponse>(request, Native.oberon_create_token);
    }

    public static CreateOberonProofResponse CreateProof(CreateOberonProofRequest request)
    {
        return Native.Call<CreateOberonProofRequest, CreateOberonProofResponse>(request, Native.oberon_create_proof);
    }

    public static VerifyOberonProofResponse VerifyProof(VerifyOberonProofRequest request)
    {
        return Native.Call<VerifyOberonProofRequest, VerifyOberonProofResponse>(request, Native.oberon_verify_proof);
    }

    public static BlindOberonTokenResponse BlindToken(BlindOberonTokenRequest request)
    {
        return Native.Call<BlindOberonTokenRequest, BlindOberonTokenResponse>(request, Native.oberon_blind_token);
    }

    public static UnBlindOberonTokenResponse UnblindToken(UnBlindOberonTokenRequest request)
    {
        return Native.Call<UnBlindOberonTokenRequest, UnBlindOberonTokenResponse>(request, Native.oberon_unblind_token);
    }
    
    public static VerifyOberonTokenResponse VerifyToken(VerifyOberonTokenRequest request)
    {
        return Native.Call<VerifyOberonTokenRequest, VerifyOberonTokenResponse>(request, Native.oberon_verify_token);
    }
}

File: java/src/main/java/trinsic/okapi/Oberon.java
package trinsic.okapi;

import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.security.v1.Security;

public class Oberon extends OkapiNative {
  public static Security.CreateOberonKeyResponse createKey(Security.CreateOberonKeyRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_create_key(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.CreateOberonKeyResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.CreateOberonTokenResponse createToken(
      Security.CreateOberonTokenRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_create_token(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.CreateOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.BlindOberonTokenResponse blindToken(
      Security.BlindOberonTokenRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_blind_token(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.BlindOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.UnBlindOberonTokenResponse unBlindToken(
      Security.UnBlindOberonTokenRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_unblind_token(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.UnBlindOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.VerifyOberonTokenResponse verifyToken(
      Security.VerifyOberonTokenRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_verify_token(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.VerifyOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.CreateOberonProofResponse createProof(
      Security.CreateOberonProofRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_create_proof(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.CreateOberonProofResponse.parseFrom(bufferToByteArray(responseBuffer));
  }

  public static Security.VerifyOberonProofResponse verifyProof(
      Security.VerifyOberonProofRequest request)
      throws DidException, InvalidProtocolBufferException {
    OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
    OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
    ExternError errBuffer = new ExternError();
    var result = getNativeLibrary().oberon_verify_proof(requestBuffer, responseBuffer, errBuffer);
    errBuffer.raiseError(result);
    return Security.VerifyOberonProofResponse.parseFrom(bufferToByteArray(responseBuffer));
  }
}
